Deduplicate the logic for resource name generation

mbaijal opened the pull request:
Deduplicate logic that calculates generated names, truncating them for certain lengths. 

Testing:
All Unit tests have passed. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

RedbackThomson commented on controllers/common_test.go:
```diff
@@ -148,4 +152,41 @@ var _ = Describe("GetGeneratedJobName", func() {
+			maxNameLen = 64
+			objectMetaName = strings.Repeat("A", 70)
+			generatedPostfix = "generated.postfix"
+			fmt.Printf("Hey Meghna, the generated name is : %s", generatedJobNameWithPostfix)
```
Another debugging mark?

RedbackThomson commented on controllers/common_test.go:
```diff
@@ -148,4 +152,41 @@ var _ = Describe("GetGeneratedJobName", func() {
+			fmt.Printf("Hey Meghna, the generated name is : %s", generatedJobNameWithPostfix)
+		})
+
+		It("Concatenates the name, shortened uid, generatedPostfix", func() {
```
nit: Update since this doesn't check it concatenates the name

RedbackThomson commented on controllers/common.go:
```diff
@@ -152,10 +152,11 @@ func Now() *metav1.Time {
-
+func GetGeneratedResourceName(objectMetaUID types.UID, objectMetaName string, maxNameLen int, generatedPostfix ...string) string {
+	uid := strings.Replace(string(objectMetaUID), "-", "", -1)
+	requiredPostfix := strings.Join(generatedPostfix, "") + uid
```
Could we instead join the requiredPostfix with a delimiter before the uid so that when we use the generatedPostfix we don't need to remember the hyphen?

For example:
`requiredPostfix := strings.Join(generatedPostfix, "") + delimiter + uid`

mbaijal replied:
> Could we instead join the requiredPostfix with a delimiter before the uid so that when we use the generatedPostfix we don't need to remember the hyphen?
> 
> For example:
> `requiredPostfix := strings.Join(generatedPostfix, "") + delimiter + uid`

I did not do this since this would require a conditional check to add the delimiter only when the generatedPostfix does exist.

mbaijal replied:
> We need to be careful of changing the format of these names. Because our installers set the image pull policy to `Always`, we risk breaking a customer setup when we change them. Let's avoid changes here unless we want to have a live-migration code.

The formats of the names are not changing here. The generated names would be the exact same as before. Can you elaborate?

cadedaniel replied:
> > We need to be careful of changing the format of these names. Because our installers set the image pull policy to Always, we risk breaking a customer setup when we change them. Let's avoid changes here unless we want to have a live-migration code.
> 
> The formats of the names are not changing here. The generated names would be the exact same as before. Can you elaborate?

I am saying that although Nick's suggestion is a good one, we shouldn't break these names until v2 unless we really need to.

RedbackThomson replied:
Any changes to these formats would necessitate a major version upgrade.

RedbackThomson commented on controllers/hosting/hostingdeployment_controller.go:
```diff
@@ -468,22 +469,28 @@ func convertTimeOrDefault(time *time.Time, defaultValue *metav1.Time) *metav1.Ti
+	k8sMaxLen := 253
+	uid := hostingDeployment.ObjectMeta.GetUID()
+	generation := strconv.FormatInt(hostingDeployment.ObjectMeta.GetGeneration(), 10)
+	generatedPostfix := generation + "-"
```
See comment on line 157 on `common.go`

mbaijal replied:
I did this intentionally, check response earlier.

RedbackThomson commented on controllers/hosting/hostingdeployment_controller_test.go:
```diff
@@ -40,6 +40,7 @@ import (
 	commonv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/common"
 	endpointconfigv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/endpointconfig"
 	hostingv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/hostingdeployment"
+	. "github.com/aws/amazon-sagemaker-operator-for-k8s/controllers"
```
Please avoid using dot imports - they give warnings in golint. I've been trying to remove our use of these in the refactor work.

RedbackThomson commented on controllers/endpointconfig/endpointconfig_controller.go:
```diff
@@ -150,7 +149,7 @@ func (r *EndpointConfigReconciler) reconcileEndpointConfig(ctx reconcileRequestC
 
 	// Get SageMaker endpointConfig description.
-	if ctx.EndpointConfigDescription, err = ctx.SageMakerClient.DescribeEndpointConfig(ctx, generateEndpointConfigName(ctx.EndpointConfig)); err != nil {
+	if ctx.EndpointConfigDescription, err = ctx.SageMakerClient.DescribeEndpointConfig(ctx, GetGeneratedResourceName(ctx.EndpointConfig.ObjectMeta.GetUID(), ctx.EndpointConfig.ObjectMeta.GetName(), 63)); err != nil {
```
Would you be able to create constants for holding the name lengths for each resource? I've included this in the TrainingJob refactor, as an example.

cadedaniel commented on controllers/model/model_controller.go:
```diff
@@ -146,7 +145,7 @@ func (r *ModelReconciler) reconcileModel(ctx reconcileRequestContext) error {
 
 	// Get SageMaker model description.
-	if ctx.ModelDescription, err = ctx.SageMakerClient.DescribeModel(ctx, generateModelName(ctx.Model)); err != nil {
+	if ctx.ModelDescription, err = ctx.SageMakerClient.DescribeModel(ctx, GetGeneratedResourceName(ctx.Model.ObjectMeta.GetUID(), ctx.Model.ObjectMeta.GetName(), 63)); err != nil {
```
Can we deduplicate the 3 identical calls to GetGeneratedResourceName?

cadedaniel commented on controllers/common_test.go:
```diff
@@ -148,4 +152,41 @@ var _ = Describe("GetGeneratedJobName", func() {
 		})
 	})
+
+	When("An additional postfix is required and maxNameLen is sufficiently large", func() {
```
Reminder for me to look at tests in-depth on second pass.

mbaijal commented on controllers/hosting/hostingdeployment_controller.go:
```diff
@@ -468,22 +469,28 @@ func convertTimeOrDefault(time *time.Time, defaultValue *metav1.Time) *metav1.Ti
+// are able to find the object.
+// Kubernetes resources can have names up to 253 characters long.
+// The characters allowed in names are: digits (0-9), lower case letters (a-z), -, and .
+func GetKubernetesNamespacedName(objectName string, hostingDeployment hostingv1.HostingDeployment) types.NamespacedName {
```
I originally added it to the common.go file but leads to a cyclic import for hostingDeployments and is not possible to do. This seemed like a better (logically common) place for the function than putting it in either model of endpoint config and then calling it in the other. Open to moving it though.

mbaijal commented on controllers/hosting/hostingdeployment_controller.go:
```diff
@@ -386,7 +387,7 @@ func (r *HostingDeploymentReconciler) createCreateEndpointInput(ctx reconcileReq
 		return nil, err
 	}
 
-	endpointName := GetSageMakerEndpointName(*ctx.Deployment)
```
@mbaijal This is pending

mbaijal commented on controllers/batchtransformjob/batchtransformjob_controller.go:
```diff
@@ -39,7 +39,12 @@ import (
+const (
+	transformResourceNotFoundApiCode string = "ValidationException"
+
+	// Defines the maximum number of characters in a SageMaker Endpoint Config Resource name
```
Fix comment

cadedaniel commented on controllers/common_test.go:
```diff
@@ -36,15 +36,25 @@ var _ = Describe("GetGeneratedJobName", func() {
+		if shouldJobNameHavePostfix {
+			generatedJobName = GetGeneratedResourceName(generatedPostfix+"-"+uidWithoutHyphens, objectMetaName, maxNameLen)
+		} else {
+			generatedJobName = GetGeneratedJobName(uid, objectMetaName, maxNameLen)
```
Why are we testing another function in `GetGeneratedResourceName` tests? This couples `GetGeneratedResourceName` with `GetGeneratedJobName`, making code harder to refactor and tests more difficult to comprehend.

I think since `GetGeneratedJobName` merely calls `GetGeneratedResourceName`, we only need a test on the calling logic for `GetGeneratedJobName`. Can we break out a new `Describe` case for `GetGeneratedJobName` which only tests 1) that the UID does not have hyphens and 2) that the output is equivalent to `GetGeneratedResourceName` output for some basic case?

cadedaniel commented on controllers/common_test.go:
```diff
@@ -148,4 +158,42 @@ var _ = Describe("GetGeneratedJobName", func() {
+		BeforeEach(func() {
+			maxNameLen = 253
+			objectMetaName = "object.meta.name"
+			generatedPostfix = "generated.postfix"
```
This usage is confusing to me. Who generates this postfix? What is it postfixed to?

cadedaniel replied:
I think these tests would be much more clear if they used `required` and `optional` strings, following the parameters of `GetGeneratedResourceName`.

mbaijal replied:
did not update the tests very well to match the changed methods, fixing

cadedaniel commented on controllers/common_test.go:
```diff
@@ -148,4 +158,42 @@ var _ = Describe("GetGeneratedJobName", func() {
+		It("Concatenates the name, shortened uid, generatedPostfix", func() {
+			Expect(generatedJobName).To(ContainSubstring(objectMetaName))
+			Expect(generatedJobName).To(ContainSubstring(uidWithoutHyphens))
+			Expect(generatedJobName).To(ContainSubstring(generatedPostfix))
```
Why are we differentiating between different substrings of the `required` parameter? Shouldn't the function's behavior be the same for the entire `required` parameter?

RedbackThomson replied:
Could we also verify the correct ordering for all the values and that they are delimited in the correct places?

mbaijal replied:
I think its still a good idea to have required = postfix+uid in order to check that if maxlen<required, its the uid part that is returned.

cadedaniel replied:
We can do the additional tests Nick suggests, but given that we don't have any requirement on the ordering I think we are good as is.

RedbackThomson commented on controllers/common_test.go:
```diff
@@ -148,4 +158,42 @@ var _ = Describe("GetGeneratedJobName", func() {
+		It("Length does not exceed maxNameLen", func() {
+			Expect(len(generatedJobName)).To(BeNumerically("<=", maxNameLen))
+		})
+	})
```
Is there a test for substrings of the required part without a required postfix? Just making sure we hit each condition

mbaijal replied:
without a required postfix - not sure what you mean. 
There is a test to ensure that if required is larger than the maxlen, uid is prioritizied.

mbaijal commented:
> Would you mind also checking to see if any of these changes introduced further linting errors? It's been a goal of mine for every PR to reduce the linting error count. I know there are still a lot, so it might be hard to see which of those have been introduced by your changes.

I have fixed all linting errors in the common.go file and a few in aws_config_loader.go (except one for which I added a TODO). The rest of the project looks clean!

cadedaniel commented on controllers/aws_config_loader.go:
```diff
@@ -28,29 +28,29 @@ import (
 )
 
-// Simple struct to facilitate loading AWS config with region- and endpoint-overrides.
+// AwsConfigLoader is a simple struct to facilitate loading AWS config with region- and endpoint-overrides.
```
Thanks again for taking care of these additional lint fixes.

cadedaniel commented on controllers/model/model_controller.go:
```diff
@@ -344,3 +334,8 @@ func (r *ModelReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
+
+// Given a Model object, generate a Sagemaker name of valid length
```
nit: `s/Sagemaker/SageMaker`.

cadedaniel commented on controllers/common_test.go:
```diff
@@ -133,19 +133,64 @@ var _ = Describe("GetGeneratedJobName", func() {
+	JustBeforeEach(func() {
+		uidWithoutHyphens = strings.Replace(string(uid), "-", "", -1)
+		generatedJobName = GetGeneratedJobName(uid, objectMetaName, maxNameLen)
+		generatedResourceName = GetGeneratedJobName(uid, objectMetaName, maxNameLen)
```
These variables are assigned the same value? I think you mean to call `GetGeneratedResourceName` with `uidWithoutHyphens`? This makes the test on 192 a very happy test that won't ever fail :)

If you are only using `generatedResourceName` in one test, it might make more sense to create it in that test alone.

mbaijal replied:
oh yes, good catch. Fixed. 

It is a little tricky to move the creation of `generatedResourceName` since the `uid` gets generated a bunch of times for each spec. 
I tried to create the `generatedResourceName` within the `Before Each` block of the test itself (inside the `When`). 
This causes the `generatedJobName` and `GetGeneratedResourceName` to be called with different `uid` and `uidWithoutHyphens`. An example of the failure log below: 

```
•••••••••••••generatedJobName is object.meta.name-bced7caf409947faa7d2161963a1bff7
generatedResourceName is : object.meta.name-
•generatedJobName is object.meta.name-3c0b1cadf1e6465d9043f791d56f8d46
generatedResourceName is : object.meta.name-bced7caf409947faa7d2161963a1bff7
•generatedJobName is object.meta.name-2e70d1f94c12493f8797310821130c58
generatedResourceName is : object.meta.name-3c0b1cadf1e6465d9043f791d56f8d46
---------------------
    Expected
        <string>: object.meta.name-2e70d1f94c12493f8797310821130c58
    to be equivalent to
        <string>: object.meta.name-3c0b1cadf1e6465d9043f791d56f8d46
```

cadedaniel replied:
Hmm.. I don't know how this issue was caused. It seems that the two `BeforeEach`s are called in the wrong order, which means that the UID from the last test and this test are used.

I wouldn't put it in a `BeforeEach` (which runs before a set of tests), and simply place it in the single test where it is used, i.e.:

```go
It("returns the same name as GetGeneratedResourceName with uidWithoutHyphens as the required string", func() {
        generatedResourceName := GetGeneratedResourceName(uidWithoutHyphens, objectMetaName, maxNameLen)
	Expect(generatedJobName).To(BeEquivalentTo(generatedResourceName))
})
```

This is pretty minor, so fine leaving as is.

cadedaniel commented on controllers/common_test.go:
```diff
@@ -133,19 +133,64 @@ var _ = Describe("GetGeneratedJobName", func() {
+			Expect(len(generatedResourceName)).To(BeNumerically("<=", maxNameLen))
+		})
+
+		It("Is contained in the full hyphen-less UID", func() {
```
nit: update to say `Is contained in the full required string`.

cadedaniel commented on controllers/common_test.go:
```diff
@@ -44,63 +43,64 @@ var _ = Describe("GetGeneratedJobName", func() {
 
 	JustBeforeEach(func() {
 		uidWithoutHyphens = strings.Replace(string(uid), "-", "", -1)
```
nit: It is weird that the `uidWithoutHyphens` is still here, since you refactored out the logic that removes hyphens. It will make test easier to understand if you remove it since readers will not have to understand this element -- the function `GetGeneratedResourceName` should work for arbitrary string input.

mbaijal replied:
It does, but the idea was to check if the length of required is more than maxLen, the second half of required which contains the uid gets picked. I changed this to not use a generated UID but a hardcoded one. Let me know if you prefer to remove this check altogether.

cadedaniel replied:
> It does, but the idea was to check if the length of required is more than maxLen, the second half of required which contains the uid gets picked

This test is fine but somewhat imprecise -- if `GetGeneratedResourceName` has room only for `required` (no `optional`), why would our tests only check that the last half of `required` is present?

To make the test better match the documented behavior of the function, we should change:
```go
It("Contains the full UID", func() {
    Expect(generatedResourceName).To(ContainSubstring(requiredPostfix))
})
```
to:
```go
It("Contains the full required string", func() {
    Expect(generatedResourceName).To(ContainSubstring(required))
})
```

The impact of this change pretty small so happy to approve either way.

cadedaniel commented:
Can we make sure to run integration tests or smoke tests on this? I think if you do a manual HPO job and manual HostingDeployment it will be a good enough test since HPO spawns TrainingJobs and HostingDeployment is the only one that uses the Kubernetes names.

mbaijal commented:
> Can we make sure to run integration tests or smoke tests on this? I think if you do a manual HPO job and manual HostingDeployment it will be a good enough test since HPO spawns TrainingJobs and HostingDeployment is the only one that uses the Kubernetes names.

I did a manual [1]hostingDeployment, [2]trainingJob, [3]BatchTransform and [4]HPO. I verified the endpoint, endpointconfig, model names for [1]. [2] and [3] did not go to completion due to incorrect specs but were successfully launched.  For [4] training jobs were successfully spawned and are in progress.
